src/Analyzers/CSharp/Analyzers/UseIsNullCheck/CSharpUseNullCheckOverTypeCheckDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,85 @@
+                   EnforceOnBuildValues.UseNullCheckOverTypeCheck,
+                   CSharpCodeStyleOptions.PreferNullCheckOverTypeCheck,
+                   CSharpAnalyzersResources.Use_is_null_check,
+                   new LocalizableResourceString(nameof(CSharpAnalyzersResources.Null_check_can_be_clarified), AnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(context =>
+            {
+                // All trees should have the same language version. Bail-out early in compilation start instead of checking every tree.
+                var tree = context.Compilation.SyntaxTrees.FirstOrDefault();
+                if (tree is null || ((CSharpParseOptions)tree.Options).LanguageVersion < LanguageVersion.CSharp9)
+                {
+                    return;
+                }
+
+                context.RegisterOperationAction(c => AnalyzeOperation(c), OperationKind.IsType, OperationKind.NegatedPattern);
+            });
+        }
+
+        private void AnalyzeOperation(OperationAnalysisContext context)
+        {
+            var option = context.Options.GetOption(CSharpCodeStyleOptions.PreferNullCheckOverTypeCheck, context.Operation.Syntax.SyntaxTree, context.CancellationToken);
+            if (!option.Value)
+            {
+                return;
+            }
+
+            if (ShouldReportDiagnostic(context.Operation))
+            {
+                var severity = option.Notification.Severity;
+                context.ReportDiagnostic(
+                    DiagnosticHelper.Create(
+                        Descriptor, context.Operation.Syntax.GetLocation(), severity, additionalLocations: null, properties: null));
+            }
+        }
+
+        private static bool ShouldReportDiagnostic(IOperation operation)
+        {
+            if (operation is IIsTypeOperation isTypeOperation)
+            {
+                // Matches 'x is MyType'
+                // isTypeOperation.TypeOperand is 'MyType'
+                // isTypeOperation.ValueOperand.Type is the type of 'x'.
+                return isTypeOperation.ValueOperand.Type is not null &&
+                    isTypeOperation.ValueOperand.Type.InheritsFromOrEquals(isTypeOperation.TypeOperand);
+            }
+            else if (operation is INegatedPatternOperation negatedPattern)
+            {
+                // Matches 'x is not MyType'
+                // InputType is the type of 'x'
+                // MatchedType is 'MyType'
+                return negatedPattern.Pattern is ITypePatternOperation typePatternOperation &&
+                    typePatternOperation.InputType.InheritsFromOrEquals(typePatternOperation.MatchedType);
```

Comment:
Shouldn't there be a check, that there is no designation present like in `is not object o`?

Reply:
@MaStr11 `negatedPattern.Pattern` won't be ITypePatternOperation for the `is not object o` case. Going to add a test.